sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,612 @@
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract ConnectionConfiguration connectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    /**
+     * Define MQTT connection configuration used to connect to the MQTT broker.
+     */
+    public Write withConnectionConfiguration(ConnectionConfiguration configuration) {
+      checkArgument(configuration != null,
+          "MqttIO.write().withConnectionConfiguration(configuration) called with null "
+              + "configuration or not called at all");
+      return builder().setConnectionConfiguration(configuration).build();
+    }
+
+    /**
+     * Define the MQTT message quality of service.
+     *
+     * <ul>
+     * <li>Quality of Service 0 (at most once) - a message won't be ack by the receiver or
+     * stored and redelivered by the sender. This is "fire and forget" and provides the same
+     * guarantee as the underlying TCP protocol.</li>
+     *
+     * <li>Quality of Service 1 (at least once) - a message will be delivered at least once to
+     * the receiver. But the message can also be delivered more than once. The sender will
+     * store the message until it gets an ack from the receiver.</li>
+     *
+     * <li>Quality of Service 2 (exactly one) - each message is received only once by the
+     * counterpart. It is the safest and also the slowest quality of service level. The
+     * guarantee is provided by two flows there and back between sender and receiver.</li>
+     * </ul>
+     *
+     * <p>If persistence is not configured, QoS 1 and 2 messages will still be delivered
+     * in the event of a network or server problem as the client will hold state in memory.
+     * If the MQTT client is shutdown or fails and persistence is not configured then
+     * delivery of QoS 1 and 2 messages can not be maintained as client-side state will
+     * be lost.
```

Comment:
I don't understand your comment: QoS 2 should work.
Let me explain.
QoS 0 means no ack (fire and forget).
QoS 1 means we wait that at least one subscriber receives the message and so send back the ack.
QoS 2 means we wait that all subscribers receive the message and have send back the ack.

So, let me take example for the two parts of the IO: Read and Write.

Let's start with the Write:
- the user defines the QoS to 2.
- the `WriteFn` use `MqttClient.publish()`. This method is blocker and wait for the completion. It means that this method is waiting for the ack from the broker. The broker will send the ack when he will receive all ack from subscriber (thanks to QoS 2).
- In the case of a failure in the `WriteFn` (for any reason), the broker won't be able to send back the ack to the publisher, and so, it won't deliver the message on the subscribers (it's dual direction mechanism).

Now, on the Read with QoS 2:
- the QoS is defined in the message.
- the Read will send the ack to the broker when the checkpoint will be finalize. The broker deals with the publisher and other subscribers depending of the QoS in the message.

I don't see problem with QoS 2.